api/v1alpha1/siteconfig_types.go

```diff
@@ -0,0 +1,461 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	bmh_v1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+
+	aiv1beta1 "github.com/openshift/assisted-service/api/v1beta1"
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// SshPrivateKeySecretRef defines the private key secret reference
+type SshPrivateKeySecretRef struct {
+	Name string `json:"name,omitempty"`
+}
+
+// MachineNetworkEntry is a single IP address block for node IP blocks.
+type MachineNetworkEntry struct {
+	// CIDR is the IP block address pool for machines within the cluster.
+	CIDR string `json:"cidr"`
+}
+
+// ClusterNetworkEntry is a single IP address block for pod IP blocks. IP blocks
+// are allocated with size 2^HostSubnetLength.
+type ClusterNetworkEntry struct {
+	// CIDR is the IP block address pool.
+	CIDR string `json:"cidr"`
+
+	// HostPrefix is the prefix size to allocate to each node from the CIDR.
+	// For example, 24 would allocate 2^8=256 adresses to each node. If this
+	// field is not used by the plugin, it can be left unset.
+	// +optional
+	HostPrefix int32 `json:"hostPrefix,omitempty"`
+}
+
+// BmcCredentialsName
+type BmcCredentialsName struct {
+	Name string `json:"name,omitempty"`
+}
+
+// IronicInspect
+type IronicInspect string
+
+// SiteConfigMap
+type SiteConfigMap struct {
+	Name string `json:"name,omitempty"`
+	// +kubebuilder:default:=ztp-site
+	Namespace string            `json:"namespace,omitempty"`
+	Data      map[string]string `json:"data,omitempty"`
+}
+
+type CrAnnotations struct {
+	Add map[string]map[string]string `json:"add,omitempty"`
```

Comment:
Also I'm not a huge fan of `Add` as it's very imperative for a declarative API. I think `Exists` would be better if you're going to ensure these annotations exist on the rendered resources.

Also how does this type work? Why not just a map rather than a map of maps?

Reply:
map of maps allows the user to specify annotations (of type `map[string]string`) for the various rendered manifest kinds. For example:
```yaml
crAnnotations:
  add:
    ClusterDeployment:
      myTestAnnotation: "success"
```

The `Add` is for indicating that the specified annotations should be “added” to existing annotations. I believe, this is to perhaps support future cases of “remove/delete” in which annotations can be removed from rendered manifests. The `CrAnnotations` here is matched to the original `SiteConfig` for backwards compatibility. @imiller0 please can you chime in on this? Can we remove the `Add` and treat the `CrAnnotations` as a `map[string]map[string]string`?

Reply:
@sakhoury my 2 cents, if we are only adding annotations to a CR and not providing any option to remove/delete, then as a default i am not for 'add' field. If it is meant to be just adding annotations, why not changing `crAnnotations` to `AddAnnotations`?

Reply:
Thanks @sabbir-47! If there is no intention to support removal of annotations, I would suggest going with `Annotations`.

Reply:
I have relabeled the field to `extraAnnotations` and have removed the `add` component.